<!DOCTYPE html><html lang="en-US"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=Edge"><title> - Archer</title><link rel="shortcut icon" href="http://localhost:4000/favicon.ico" type="image/x-icon"><link rel="stylesheet" href="http://localhost:4000/assets/css/just-the-docs.css"> <script type="text/javascript" src="http://localhost:4000/assets/js/vendor/lunr.min.js"></script> <script type="text/javascript" src="http://localhost:4000/assets/js/just-the-docs.js"></script><meta name="viewport" content="width=device-width, initial-scale=1"><title>Archer | PhD repository</title><meta name="generator" content="Jekyll v3.8.5" /><meta property="og:title" content="Archer" /><meta property="og:locale" content="en_US" /><meta name="description" content="PhD repository" /><meta property="og:description" content="PhD repository" /><link rel="canonical" href="http://localhost:4000/docs/circuitSNP/code/snp_vec_creation/" /><meta property="og:url" content="http://localhost:4000/docs/circuitSNP/code/snp_vec_creation/" /><meta property="og:site_name" content="Archer" /> <script type="application/ld+json"> {"@type":"WebPage","url":"http://localhost:4000/docs/circuitSNP/code/snp_vec_creation/","headline":"Archer","description":"PhD repository","@context":"https://schema.org"}</script><body> <svg xmlns="http://www.w3.org/2000/svg" style="display: none;"> <symbol id="link" viewBox="0 0 16 16"><title>Link</title><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path> </symbol> </svg><div class="page-wrap"><div class="side-bar"><div class="site-header"> <a href="http://localhost:4000" class="site-title lh-tight"> Archer </a> <button class="menu-button fs-3 js-main-nav-trigger" data-text-toggle="Hide" type="button">Menu</button></div><div class="navigation main-nav js-main-nav"><nav role="navigation" aria-label="Main navigation"><ul class="navigation-list"><li class="navigation-list-item active"><a href="http://localhost:4000/docs/circuitSNP/code/6_split_data_for_denny_NN/" class="navigation-list-link"></a><li class="navigation-list-item active"><a href="http://localhost:4000/docs/circuitSNP/code/snp_vec_creation/" class="navigation-list-link active"></a><li class="navigation-list-item"><a href="http://localhost:4000/docs/circuitSNP/code/split_data/" class="navigation-list-link">Now we split data</a><li class="navigation-list-item"><a href="http://localhost:4000/docs/github%20misc/github/" class="navigation-list-link">Github doc</a><li class="navigation-list-item active"><a href="http://localhost:4000/docs/hpc/copying_data_to_hpc/" class="navigation-list-link"></a><li class="navigation-list-item"><a href="http://localhost:4000/" class="navigation-list-link">Home</a><li class="navigation-list-item"><a href="http://localhost:4000/docs/circuitSNP/circuitSNP/" class="navigation-list-link">circuitSNP</a><ul class="navigation-list-child-list "><li class="navigation-list-item "><a href="http://localhost:4000/docs/circuitSNP/code/code_documentation/" class="navigation-list-link">Python Implementation</a><ul class="navigation-list-child-list"><li class="navigation-list-item "> <a href="http://localhost:4000/docs/circuitSNP/code/1_preliminary_setup/" class="navigation-list-link">1. Environment Configuration</a><li class="navigation-list-item "> <a href="http://localhost:4000/docs/circuitSNP/code/2_download_data/" class="navigation-list-link">2. Download data</a><li class="navigation-list-item "> <a href="http://localhost:4000/docs/circuitSNP/code/3_analyse_lcl_chromatin/" class="navigation-list-link">3. Detect open / closed regions</a><li class="navigation-list-item "> <a href="http://localhost:4000/docs/circuitSNP/code/4_analyse_motif_window_membership/" class="navigation-list-link">4. Motif membership</a><li class="navigation-list-item "> <a href="http://localhost:4000/docs/circuitSNP/code/5_generate_T_matrix/" class="navigation-list-link">5. Generate T matrix</a><li class="navigation-list-item "> <a href="http://localhost:4000/docs/circuitSNP/code/7_run_NN/" class="navigation-list-link">6. Run NN</a><li class="navigation-list-item "> <a href="http://localhost:4000/docs/circuitSNP/code/functions/" class="navigation-list-link">Imports and functions</a></ul></ul><li class="navigation-list-item"><a href="http://localhost:4000/docs/hpc/spartan/" class="navigation-list-link">Spartan HPC Usage</a><ul class="navigation-list-child-list "><li class="navigation-list-item "><a href="http://localhost:4000/docs/hpc/slurm/" class="navigation-list-link">Slurm example script</a></ul><li class="navigation-list-item"><a href="http://localhost:4000/docs/jekyll_maintenance/jekyll_setup/" class="navigation-list-link">Web Development</a></ul></nav></div><footer class="site-footer"><p class="text-small text-grey-dk-000 mb-4">This site uses <a href="https://github.com/pmarsceill/just-the-docs">Just the Docs</a>, a documentation theme for Jekyll.</p></footer></div><div class="main-content-wrap js-main-content" tabindex="0"><div class="main-content"><div class="page-header js-page-header"><div class="search"><div class="search-input-wrap"> <input type="text" class="js-search-input search-input" tabindex="0" placeholder="Search Archer" aria-label="Search Archer" autocomplete="off"> <svg width="14" height="14" viewBox="0 0 28 28" xmlns="http://www.w3.org/2000/svg" class="search-icon"><title>Search</title><g fill-rule="nonzero"><path d="M17.332 20.735c-5.537 0-10-4.6-10-10.247 0-5.646 4.463-10.247 10-10.247 5.536 0 10 4.601 10 10.247s-4.464 10.247-10 10.247zm0-4c3.3 0 6-2.783 6-6.247 0-3.463-2.7-6.247-6-6.247s-6 2.784-6 6.247c0 3.464 2.7 6.247 6 6.247z"/><path d="M11.672 13.791L.192 25.271 3.02 28.1 14.5 16.62z"/></g></svg></div><div class="js-search-results search-results-wrap"></div></div></div><div class="page"><div id="main-content" class="page-content" role="main"><p>#####generate_SNP_vectors_6.py</p><p>In this stage, we wish to identify all SNPs in a region. We want to generate 3 outputs:</p><p>The aim:</p><ol><li>Generate F vectors (both ref and alternate contained)<li>Generate reference vectors<li>Generate alternate vectors</ol><p>Dependencies, functions, imports</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">from</span> <span class="nn">functions</span> <span class="kn">import</span> <span class="o">*</span>
</code></pre></div></div><p>Data downloaded in earlier steps:</p><p>SNP_READS is all SNPs</p><p>dsqtl consists of SNPs which are also dsQTLs</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#read in SNPs
</span><span class="n">SNP_READS</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">DOWN_DIR</span>  <span class="o">+</span> <span class="s">'compendium/all_compendium_reads.csv'</span><span class="p">)</span>
<span class="n">SNP_READS</span><span class="p">[</span><span class="s">'SNP_window_index'</span><span class="p">]</span><span class="o">=-</span><span class="mi">1</span>

<span class="c1">#read in dsQTLs
</span><span class="n">dsqtl</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">DOWN_DIR</span>  <span class="o">+</span> <span class="s">'test/dsQTL.eval.txt'</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
<span class="n">dsqtl</span><span class="p">[</span><span class="s">'SNP_window_index'</span><span class="p">]</span><span class="o">=-</span><span class="mi">1</span>

</code></pre></div></div><p>rename for easier referencing</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>
<span class="n">SNP_READS</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s">'pos    (0-based)'</span><span class="p">:</span><span class="s">'snp_start'</span><span class="p">},</span> <span class="n">inplace</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span> 
<span class="n">SNP_READS</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s">'pos1   (1-based)'</span><span class="p">:</span><span class="s">'snp_end'</span><span class="p">},</span> <span class="n">inplace</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span> 
<span class="n">SNP_READS</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s">'motif  (motif ID, see factorNames.txt)'</span><span class="p">:</span><span class="s">'motif'</span><span class="p">},</span> <span class="n">inplace</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span> 

<span class="n">dsqtl</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s">'pos0'</span><span class="p">:</span><span class="s">'snp_start'</span><span class="p">},</span> <span class="n">inplace</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span> 
<span class="n">dsqtl</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s">'pos'</span><span class="p">:</span><span class="s">'snp_end'</span><span class="p">},</span> <span class="n">inplace</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)</span> 

</code></pre></div></div><p>before manipulating tables, number of rows in each is</p><div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&gt;&gt;&gt; SNP_READS.shape
(5202041, 12)
&gt;&gt;&gt; dsqtl.shape
(27702, 24)
</code></pre></div></div><p>remove sex chromosomes</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">SNP_READS</span><span class="o">=</span><span class="n">SNP_READS</span><span class="p">[</span><span class="o">~</span><span class="n">SNP_READS</span><span class="o">.</span><span class="nb">chr</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s">'chrX'</span><span class="p">,</span><span class="s">'chrY'</span><span class="p">])]</span>
<span class="n">dsqtl</span><span class="o">=</span><span class="n">dsqtl</span><span class="p">[</span><span class="o">~</span><span class="n">dsqtl</span><span class="o">.</span><span class="nb">chr</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s">'chrX'</span><span class="p">,</span><span class="s">'chrY'</span><span class="p">])]</span>
</code></pre></div></div><p>new number of rows in each</p><div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&gt;&gt;&gt; SNP_READS.shape
(5057359, 12)
&gt;&gt;&gt; dsqtl.shape
(27702, 24)
</code></pre></div></div><p>Number of unique SNP is 3766427</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">&gt;&gt;&gt;</span> <span class="n">SNP_READS</span><span class="p">[[</span><span class="s">'chr'</span><span class="p">,</span><span class="s">'snp_start'</span><span class="p">,</span><span class="s">'snp_end'</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span>
<span class="p">(</span><span class="mi">3766427</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
</code></pre></div></div><p>But we should expect this number to be larger when we consider that duplicate SNP (same location and chromosome) will overlap with different SNPs</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">SNP_READS</span><span class="p">[[</span><span class="s">'chr'</span><span class="p">,</span><span class="s">'snp_start'</span><span class="p">,</span><span class="s">'snp_end'</span><span class="p">,</span><span class="s">'motif'</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span>
<span class="p">(</span><span class="mi">5057359</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
</code></pre></div></div><p>So exactly same as the number of rows in entire table</p><p>Number of unique dsQTL in file is</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">&gt;&gt;&gt;</span> <span class="n">dsqtl</span><span class="p">[[</span><span class="s">'chr'</span><span class="p">,</span><span class="s">'snp_start'</span><span class="p">,</span><span class="s">'snp_end'</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span>
<span class="p">(</span><span class="mi">27702</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
</code></pre></div></div><p>we want to find out how many dsQTL exist in SNP_READS (or how many SNP are dsQTL)</p><p>can use merge function</p><p>note that set indicator=True gives extra column <code class="language-plaintext highlighter-rouge">._merge</code> that counts outcome for particular row - whether row is common to both data frames, or only one</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">common_dsqtl_snp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">dsqtl</span><span class="p">,</span><span class="n">SNP_READS</span><span class="p">[[</span><span class="s">'chr'</span><span class="p">,</span><span class="s">'snp_start'</span><span class="p">,</span><span class="s">'motif'</span><span class="p">]],</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s">'chr'</span><span class="p">,</span><span class="s">'snp_start'</span><span class="p">],</span><span class="n">how</span><span class="o">=</span><span class="s">'left'</span><span class="p">,</span><span class="n">indicator</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="c1">#we are interested in the number of items that matched in both (indicates dsqtl in all SNP)
</span><span class="n">common_dsqtl_snp</span><span class="o">.</span><span class="n">_merge</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>

</code></pre></div></div><p>left_only indicates values that only exist in dsqtl table so number of SNP exist in both is 8898 (shown below)</p><div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">&gt;&gt;&gt;</span> common_dsqtl_snp._merge.value_counts<span class="o">()</span>
left_only     21849
both           8898
right_only        0
Name: _merge, dtype: int64
</code></pre></div></div><p>verify this with a right join should get 8898 again</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">common_dsqtl_snp_right</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">dsqtl</span><span class="p">,</span><span class="n">SNP_READS</span><span class="p">[[</span><span class="s">'chr'</span><span class="p">,</span><span class="s">'snp_start'</span><span class="p">,</span><span class="s">'motif'</span><span class="p">]],</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s">'chr'</span><span class="p">,</span><span class="s">'snp_start'</span><span class="p">],</span><span class="n">how</span><span class="o">=</span><span class="s">'right'</span><span class="p">,</span><span class="n">indicator</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="c1">#we are interested in the number of items that matched in both (indicates dsqtl in all SNP)
</span><span class="n">common_dsqtl_snp_right</span><span class="o">.</span><span class="n">_merge</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</code></pre></div></div><p>perfect - therefore we have 8898 SNP which are also dsQTL in SNP_READS</p><div class="language-sh highlighter-rouge"><div class="highlight"><pre class="highlight"><code>... common_dsqtl_snp_right._merge.value_counts<span class="o">()</span>
right_only    5048461
both             8898
left_only           0
</code></pre></div></div><p>read in locations of motifs</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">relevant_motifs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_feather</span><span class="p">(</span><span class="n">COMP_DIR</span> <span class="o">+</span> <span class="s">'2_motifs_in_lcl_windows.feather'</span><span class="p">)</span>
<span class="o">&gt;&gt;&gt;</span> <span class="n">relevant_motifs</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
   <span class="n">index</span> <span class="n">chrom</span>    <span class="n">start</span>     <span class="n">stop</span>   <span class="n">motif</span>       <span class="nb">float</span> <span class="n">sign</span>  <span class="n">chrom_dict</span>  <span class="n">d_motif_membership</span>
<span class="mi">0</span>     <span class="mi">21</span>  <span class="n">chr1</span>  <span class="mi">2203656</span>  <span class="mi">2203666</span>  <span class="n">M00140</span>  <span class="mf">326.956217</span>   <span class="o">+</span>           <span class="mi">70</span>                 <span class="mi">918</span>
<span class="mi">1</span>     <span class="mi">38</span>  <span class="n">chr1</span>  <span class="mi">4815881</span>  <span class="mi">4815891</span>  <span class="n">M00140</span>  <span class="mf">111.470794</span>   <span class="o">+</span>           <span class="mi">70</span>                <span class="mi">3329</span>
<span class="mi">2</span>     <span class="mi">42</span>  <span class="n">chr1</span>  <span class="mi">5034577</span>  <span class="mi">5034587</span>  <span class="n">M00140</span>    <span class="mf">8.410601</span>   <span class="o">-</span>           <span class="mi">70</span>                <span class="mi">3579</span>
<span class="mi">3</span>     <span class="mi">50</span>  <span class="n">chr1</span>  <span class="mi">5238635</span>  <span class="mi">5238645</span>  <span class="n">M00140</span>    <span class="mf">4.346002</span>   <span class="o">-</span>           <span class="mi">70</span>                <span class="mi">3783</span>

</code></pre></div></div><p>convert chromosme to int for to use in CUDA func</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>

<span class="c1">#convert chromsome to number...
</span><span class="n">all_chromosomes</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">SNP_READS</span><span class="o">.</span><span class="nb">chr</span><span class="p">))</span>
<span class="n">chr_dict</span><span class="o">=</span><span class="p">{}</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">all_chromosomes</span><span class="p">):</span>
	<span class="n">chr_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]</span><span class="o">=</span><span class="n">i</span>

<span class="c1">#convert chromosome to int
</span><span class="n">SNP_READS</span><span class="p">[</span><span class="s">'chr_int'</span><span class="p">]</span><span class="o">=-</span><span class="mi">1</span>
<span class="n">SNP_READS</span><span class="o">=</span><span class="n">SNP_READS</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">chr_int</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s">'chr'</span><span class="p">]</span><span class="o">.</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="nb">chr</span><span class="p">:</span> <span class="n">chr_dict</span><span class="p">[</span><span class="nb">chr</span><span class="p">]))</span>

<span class="n">dsqtl</span><span class="p">[</span><span class="s">'chr_int'</span><span class="p">]</span><span class="o">=-</span><span class="mi">1</span>
<span class="n">dsqtl</span><span class="o">=</span><span class="n">dsqtl</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">chr_int</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s">'chr'</span><span class="p">]</span><span class="o">.</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="nb">chr</span><span class="p">:</span> <span class="n">chr_dict</span><span class="p">[</span><span class="nb">chr</span><span class="p">]))</span>

</code></pre></div></div><p>convert motif to int to use in CUDA func</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>
<span class="c1">#convert motif to int
</span><span class="kn">from</span> <span class="nn">functions</span> <span class="kn">import</span> <span class="o">*</span>
<span class="n">SNP_READS</span><span class="p">[</span><span class="s">'motif_int'</span><span class="p">]</span><span class="o">=</span><span class="n">SNP_READS</span><span class="p">[</span><span class="s">'motif'</span><span class="p">]</span><span class="o">.</span><span class="nb">map</span><span class="p">(</span><span class="n">motif_dict</span><span class="p">)</span>



</code></pre></div></div><p>subset to motif only under consideration</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#we only want to select motif &lt;= M1372
</span><span class="n">SNP_READS</span><span class="p">[</span><span class="n">SNP_READS</span><span class="o">.</span><span class="n">motif_int</span><span class="o">&lt;=</span><span class="mi">1372</span><span class="p">]</span>

<span class="n">SNP_READS</span><span class="o">=</span><span class="n">SNP_READS</span><span class="p">[</span><span class="n">SNP_READS</span><span class="o">.</span><span class="n">motif_int</span><span class="o">&lt;=</span><span class="mi">1372</span><span class="p">]</span>
</code></pre></div></div><p>double check no NAN vals</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">&gt;&gt;&gt;</span> <span class="n">SNP_READS</span><span class="o">.</span><span class="n">motif_int</span><span class="o">.</span><span class="nb">apply</span><span class="p">(</span><span class="n">isnan</span><span class="p">)</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
<span class="bp">False</span>    <span class="mi">3134652</span>
<span class="n">Name</span><span class="p">:</span> <span class="n">motif_int</span><span class="p">,</span> <span class="n">dtype</span><span class="p">:</span> <span class="n">int64</span>
</code></pre></div></div><p>So all values in motif_int are not NAN, so all are from motifs in 1-1372</p><p>Now see how many unique SNP. We can use snp_start as all starts correspond uniquely to each individual SNP</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">SNP_READS</span><span class="p">[[</span><span class="s">'chr_int'</span><span class="p">,</span><span class="s">'snp_start'</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span>
<span class="p">(</span><span class="mi">2388566</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
</code></pre></div></div><p>So we have 2388566 individual SNP</p><p>How many of those are dsQTL - 4165</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">unique_SNP</span><span class="o">=</span><span class="n">SNP_READS</span><span class="p">[[</span><span class="s">'chr_int'</span><span class="p">,</span><span class="s">'snp_start'</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<span class="n">reduced_dsqtl_snp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">dsqtl</span><span class="p">,</span><span class="n">unique_SNP</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s">'chr_int'</span><span class="p">,</span><span class="s">'snp_start'</span><span class="p">],</span><span class="n">how</span><span class="o">=</span><span class="s">'left'</span><span class="p">,</span><span class="n">indicator</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>

<span class="n">reduced_dsqtl_snp</span><span class="o">.</span><span class="n">_merge</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
<span class="n">left_only</span>     <span class="mi">23537</span>
<span class="n">both</span>           <span class="mi">4165</span>
<span class="n">right_only</span>        <span class="mi">0</span>
</code></pre></div></div><p>re-merge to map unique_SNP back</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">SNP_READS</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">SNP_READS</span><span class="p">,</span><span class="n">unique_SNP</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s">'chr_int'</span><span class="p">,</span><span class="s">'snp_start'</span><span class="p">],</span><span class="n">how</span><span class="o">=</span><span class="s">'left'</span><span class="p">,</span><span class="n">indicator</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</code></pre></div></div><p>check that all match up</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="o">&gt;&gt;&gt;</span> <span class="n">SNP_READS</span><span class="o">.</span><span class="n">_merge</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
<span class="n">both</span>          <span class="mi">3134652</span>
<span class="n">right_only</span>          <span class="mi">0</span>
<span class="n">left_only</span>           <span class="mi">0</span>
<span class="n">Name</span><span class="p">:</span> <span class="n">_merge</span><span class="p">,</span> <span class="n">dtype</span><span class="p">:</span> <span class="n">int64</span>
</code></pre></div></div><p>move data onto CUDA array</p><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">in_snp_index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">SNP_READS</span><span class="o">.</span><span class="n">SNP_INDEX</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>

<span class="n">in_snp_effect</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">SNP_READS</span><span class="o">.</span><span class="n">effect</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>

<span class="n">in_snp_motif</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">SNP_READS</span><span class="o">.</span><span class="n">motif_int</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>

<span class="n">in_snp_priorlodds</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">SNP_READS</span><span class="o">.</span><span class="n">ref_priorlodds</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>

<span class="n">in_snp_altlodds</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">SNP_READS</span><span class="o">.</span><span class="n">alt_priorlodds</span><span class="p">,</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>

<span class="n">in_all_motifs</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1373</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
</code></pre></div></div><div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1">#T matrix constructed from number of unique snp, and number of motif
</span><span class="n">T_mat</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">([</span><span class="n">unique_SNP</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">NUMBER_OF_MOTIFS</span><span class="p">],</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="nb">bool</span><span class="p">)</span>

<span class="c1">#now generate placeholders for output matrices on CUDA memory
</span><span class="n">out_T_matrix_F</span> <span class="o">=</span> <span class="n">cuda</span><span class="o">.</span><span class="n">device_array_like</span><span class="p">(</span><span class="n">T_mat</span><span class="p">)</span>
<span class="n">out_T_matrix_Ref</span> <span class="o">=</span> <span class="n">cuda</span><span class="o">.</span><span class="n">device_array_like</span><span class="p">(</span><span class="n">T_mat</span><span class="p">)</span>
<span class="n">out_T_matrix_Alt</span> <span class="o">=</span> <span class="n">cuda</span><span class="o">.</span><span class="n">device_array_like</span><span class="p">(</span><span class="n">T_mat</span><span class="p">)</span>
<span class="c1">#run function
</span></code></pre></div></div></div></div></div></div>
